django/contrib/postgres/operations.py

```diff
@@ -164,3 +164,65 @@ def database_backwards(self, app_label, schema_editor, from_state, to_state):
             to_model_state = to_state.models[app_label, self.model_name_lower]
             index = to_model_state.get_index_by_name(self.name)
             schema_editor.add_index(model, index, concurrently=True)
+
+
+class CreateCollation(Operation):
+    reversible = True
+
+    def __init__(self, name, *, locale=None, provider=None, deterministic=True):
```

Comment:
A couple of things here...

Django 3.2 will still support PostgreSQL 9.6 according to the [wiki](https://code.djangoproject.com/wiki/SupportedDatabaseVersions). Specifying the `provider` was not supported until PostgreSQL 10. I suggest adding something like the following in `.database_forwards()`:

```python
        if (
            self.provider is not None and
            not schema_editor.connection.features.supports_alternate_collation_providers
        ):
            raise NotSupportedError('Specifying a collation provider requires PostgreSQL 10+.')
```

I'm also not sure whether we want to validate that the value of `provider` is one of `'icu'` or `'libc'`. I'll defer to a fellow on that.

I'm not sure if there is a case for supporting `lc_collate` and `lc_type`? If so, we'd just need to ensure that it is those mutually exclusive with `locale`. I don't know how useful that would be in practice, but thought I'd ask.

Other than that I'm happy that we steer clear of supporting `version` or copying from an existing collation. They sound like far too much extra complexity.

Reply:
Ah I didn't know about that in 9.6, I will add that check. I don't really want to validate the provider on the off-chance that Postgres adds more, I don't want to have to come back just to add them, and the database gives a good error message if you try to use a non-existent provider: `django.db.utils.ProgrammingError: unrecognized collation provider: blah` - seems okay to me.

As for `lc_collate` and `lc_type`... yeah I thought about it, but to me it just seemed like extra code for no real benefit. I'm open to changing my mind on that if someone can provide a compelling reason, though.

Reply:
That's all fair. I'm happy with just adding the check for 10+.
I just like to ask all the questions to ensure that we think of everything. :wink: